cfgov/research_reports/models.py

```diff
@@ -0,0 +1,235 @@
+import os
+
+from django.conf import settings
+from django.core import management
+from django.db import models
+from django.utils.safestring import mark_safe
+
+from wagtail.admin.edit_handlers import (
+    FieldPanel, InlinePanel, MultiFieldPanel, ObjectList, TabbedInterface
+)
+from wagtail.admin.forms.models import WagtailAdminModelFormMetaclass
+from wagtail.admin.forms.pages import WagtailAdminPageForm
+from wagtail.core.models import PageManager
+from wagtail.documents.edit_handlers import DocumentChooserPanel
+
+from modelcluster.fields import ParentalKey
+from modelcluster.models import ClusterableModel
+
+from v1.models import SublandingFilterablePage
+from v1.models.base import CFGOVPage
+
+
+alpha_map = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+
+
+def get_report_parts(is_appendix=False):
+    def format(s, *args):
+        if(is_appendix):
+            return s.format(*[alpha_map[arg] for arg in args])
+        else:
+            return s.format(*[arg + 1 for arg in args])
+
+    def sec(request, page):
+        sections = [section for section in
+                    page.report_sections.all().order_by('pk')
+                    if section.is_appendix == is_appendix]
+
+        return [{
+            'expanded': False,
+            'title': section.header,
+            'body': section.body,
+            'url': format('#section-{}', i),
+            'numbering': format('{}: ' if is_appendix else '{}. ', i),
+            'children': [{
+                'title': subsection.header,
+                'body': subsection.body,
+                'url': format('#section-{}.{}', i, j),
+                'numbering': '' if is_appendix else format('{}.{} ', i, j),
+                'children': [{
+                    'title': section3.header,
+                    'body': section3.body,
+                    'url': format('#section-{}.{}.{}', i, j, k),
+                    'numbering': '' if is_appendix else
+                                 format('{}.{}.{} ', i, j, k),
+                } for k, section3 in enumerate(
+                    subsection.report_sections_level_three.all().order_by('pk')
+                )]
+            } for j, subsection in enumerate(
+                section.report_subsections.all().order_by('pk')
+            )]
+        } for i, section in enumerate(sections)]
+
+    return sec
+
+
+get_report_sections = get_report_parts()
+get_report_appendices = get_report_parts(True)
+
+
+def get_researchers():
+    return dict([
+        (r.title, r.url) for r in
+        SublandingFilterablePage.objects.get(pk=4833).get_children()
+    ])
```

Comment:
Can we find some other way to do this that doesn't rely on a hardcoded page ID that only exists in production dumps? For example, what about doing a bulk lookup of DocumentDetailPages by researcher name? Is it possible for pages to have these names but exist elsewhere in the tree? It'd be nice if we had a ResearcherPage to help limit this query.

Reply:
Yeah this logic could likely be modified to not target the researchers parent page directly, as you intimated above, this was mostly done to limit the query.

Reply:
Fixed in df7ee34e64732c7e48a8a3bbb838e6cd5d4372d9